### PR TITLE
VEIOSS-152 Implement Bank Details Section

### PR DESCRIPTION
**Pull request by pbyrne84**
image

image

image

image

### CODE REVIEW

File: app/validation/Validation.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package validation
+
+object Validation {
+  val bankAccountNamePattern = """^[A-Za-z0-9\/\-?:().,'+ ]*$"""
```
Comment (pbyrne84) [inline app/validation/Validation.scala:20]: Regex copied from 
https://github.com/hmrc/one-stop-shop-registration-frontend/blob/main/app/forms/Validation/Validation.scala#L10-L10

File: app/controllers/actions/AuthenticatedControllerComponents.scala [hunk 2]
```diff
@@ -52,6 +52,9 @@ trait AuthenticatedControllerComponents extends MessagesControllerComponents {
       getData
   }
 
+  def authAndGetDataAndCheckVerifyEmail(): ActionBuilder[AuthenticatedDataRequest, AnyContent] =
```
Comment (pbyrne84) [inline app/controllers/actions/AuthenticatedControllerComponents.scala:55]: Not sure what to do here. Logic will be done in a later story

File: test/models/BankDetailsSpec.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import org.scalatest.matchers.must.Matchers
+import play.api.libs.json.Json
+
+class BankDetailsSpec extends AnyFreeSpec with Matchers with EitherValues {
+
+  "BankDetails" - {
+
+    "must exclude trailing and leading whitespace and double spaces" in {
+      val ibanText = "GB33BUKB20201555555555"
+      val bankDetails = Iban(ibanText).map { iban =>
```
Comment (pbyrne84) [inline test/models/BankDetailsSpec.scala:30]: Switch to .map().value as the original version used matching and had incomplete match warnings.

File: app/forms/validation/Validation.scala [hunk 4]
```diff
@@ -20,4 +20,5 @@ object Validation {
   val commonTextPattern = """^(?!^[’'"])(?:[A-Za-z0-9À-ÿ \!\)\(.,_/’'"&-]|[’'"](?=[A-Za-z0-9À-ÿ \!\)\(.,_/’'"&-]))*[A-Za-z0-9À-ÿ \!\)\(.,_/’'"&-](?<![’'"]$)$"""
   val emailPattern = """^(?:[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*\")@(?:(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-zA-Z0-9-]*[a-zA-Z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])$"""
   val telephonePattern = """^\+[0-9 ]{1,18}$|^[0-9 ]{1,19}$"""
+  val bankAccountNamePattern = """^[A-Za-z0-9\/\-?:().,'+ ]*$"""
```
Comment (pbyrne84) [inline app/forms/validation/Validation.scala:23]: Copied from 
https://github.com/hmrc/one-stop-shop-registration-frontend/blob/4fb8e6f7d1fa10b85746e9af63e6528858326943/app/forms/Validation/Validation.scala#L24

File: app/controllers/actions/AuthenticatedControllerComponents.scala (hunk 2 above)
Comment (warrenskiwork1979) [reply app/controllers/actions/AuthenticatedControllerComponents.scala:55]: Yeah that'll be implemented in the email verification ticket so you can remove this for now.

File: app/controllers/BankDetailsController.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.BankDetailsView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class BankDetailsController @Inject()(override val messagesApi: MessagesApi,
+                            controllerComponents: AuthenticatedControllerComponents,
```
Comment (warrenskiwork1979) [inline app/controllers/BankDetailsController.scala:31]: For consistency can we rename this cc?

File: app/forms/validation/Validation.scala (hunk 4 above)
Comment (warrenskiwork1979) [reply app/forms/validation/Validation.scala:23]: That's fine, the majority of this service will be taken from OSS with the exception of navigation.

**Review by warrenskiwork1979 — COMMENTED** [review]
Just a couple of comments, nothing major

**Review by warrenskiwork1979 — APPROVED** [review]

File: conf/app.routes [hunk 6]
```diff
@@ -111,3 +111,8 @@ POST        /remove-uk-trading-name/:index                       controllers.tra
 
 GET         /remove-all-trading-names                            controllers.tradingNames.DeleteAllTradingNamesController.onPageLoad(waypoints: Waypoints ?= EmptyWaypoints)
 POST        /remove-all-trading-names                            controllers.tradingNames.DeleteAllTradingNamesController.onSubmit(waypoints: Waypoints ?= EmptyWaypoints)
+
+GET         /bank-details                                        controllers.BankDetailsController.onPageLoad(waypoints: Waypoints ?= EmptyWaypoints)
+POST        /bank-details                                        controllers.BankDetailsController.onSubmit(waypoints: Waypoints ?= EmptyWaypoints)
+GET         /check-bank-details                                  controllers.BankDetailsController.onPageLoad(waypoints: Waypoints ?= EmptyWaypoints)
+POST        /check-bank-details                                  controllers.BankDetailsController.onSubmit(waypoints: Waypoints ?= EmptyWaypoints)
```
Comment (hmrcvmu) [inline conf/app.routes:118]: Do we need the `/check-bank-details` routes ?

Comment (warrenskiwork1979) [reply conf/app.routes:118]: No they can be removed.

**Review by hmrcvmu — DISMISSED** [review]

**Review by warrenskiwork1979 — APPROVED** [review]

**Pull request merged**